File: libgdx_example/src/main/java/com/project/GameApp.java
package com.project;

import com.badlogic.gdx.Game;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.JsonReader;
import com.badlogic.gdx.utils.JsonValue;
import com.badlogic.gdx.utils.ObjectMap;
import com.badlogic.gdx.utils.ObjectSet;

public class GameApp extends Game {

    private final AssetManager assetManager = new AssetManager();
    private final Array<String> menuOptions = new Array<>();
    private final Array<String> levelNames = new Array<>();
    private final Array<Array<String>> referencedImageFilesByLevel = new Array<>();
    private final ObjectSet<String> queuedAssets = new ObjectSet<>();
    private final ObjectMap<String, String> animationMediaById = new ObjectMap<>();

    private SpriteBatch batch;
    private ShapeRenderer shapeRenderer;
    private BitmapFont font;

    @Override
    public void create() {
        batch = new SpriteBatch();
        shapeRenderer = new ShapeRenderer();
        font = new BitmapFont();
        font.getData().markupEnabled = false;
        loadProjectData();
        setScreen(new MenuScreen(this));
    }

    public SpriteBatch getBatch() {
        return batch;
    }

    public ShapeRenderer getShapeRenderer() {
        return shapeRenderer;
    }

    public BitmapFont getFont() {
        return font;
    }

    public AssetManager getAssetManager() {
        return assetManager;
    }

    public Array<String> getMenuOptions() {
        return menuOptions;
    }

    public String getLevelName(int levelIndex) {
        if (levelIndex < 0 || levelIndex >= levelNames.size) {
            return "Unknown";
        }
        return levelNames.get(levelIndex);
    }

    public void queueReferencedAssetsForLevel(int levelIndex) {
        if (levelIndex < 0 || levelIndex >= referencedImageFilesByLevel.size) {
            return;
        }

        Array<String> levelFiles = referencedImageFilesByLevel.get(levelIndex);
        for (String relativePath : levelFiles) {
            String assetPath = "levels/" + relativePath;
            if (!queuedAssets.contains(assetPath)) {
                assetManager.load(assetPath, Texture.class);
                queuedAssets.add(assetPath);
            }
        }
    }

    public void unloadReferencedAssetsForLevel(int levelIndex) {
        if (levelIndex < 0 || levelIndex >= referencedImageFilesByLevel.size) {
            return;
        }

        Array<String> levelFiles = referencedImageFilesByLevel.get(levelIndex);
        for (String relativePath : levelFiles) {
            String assetPath = "levels/" + relativePath;
            if (assetManager.isLoaded(assetPath, Texture.class)) {
                assetManager.unload(assetPath);
            }
            queuedAssets.remove(assetPath);
        }
    }

    private void loadProjectData() {
        menuOptions.clear();
        levelNames.clear();
        referencedImageFilesByLevel.clear();
        animationMediaById.clear();

        FileHandle gameDataFile = Gdx.files.internal("levels/game_data.json");
        if (!gameDataFile.exists()) {
            addFallbackLevels();
            return;
        }

        try {
            JsonValue root = new JsonReader().parse(gameDataFile);
            JsonValue levels = root.get("levels");
            if (levels == null || !levels.isArray() || levels.size <= 0) {
                addFallbackLevels();
                return;
            }

            loadAnimationsFile(root);

            int index = 0;
            for (JsonValue level = levels.child; level != null; level = level.next) {
                String levelName = level.getString("name", "Level " + index);
                levelNames.add(levelName);
                menuOptions.add("LEVEL " + index);
                Array<String> levelImageFiles = new Array<>();
                collectImageFiles(level, levelImageFiles);
                collectAnimationMediaForLevel(level, levelImageFiles);
                referencedImageFilesByLevel.add(levelImageFiles);
                index++;
            }
        } catch (Exception ex) {
            Gdx.app.error("GameApp", "Failed to parse levels/game_data.json", ex);
            addFallbackLevels();
        }
    }

    private void addFallbackLevels() {
        levelNames.add("Level 0");
        levelNames.add("Level 1");
        menuOptions.add("LEVEL 0");
        menuOptions.add("LEVEL 1");
        referencedImageFilesByLevel.add(new Array<>());
        referencedImageFilesByLevel.add(new Array<>());
    }

    private void loadAnimationsFile(JsonValue root) {
        String animationsFilePath = root.getString("animationsFile", null);
        if (animationsFilePath == null || animationsFilePath.isEmpty()) {
            return;
        }

        FileHandle file = Gdx.files.internal("levels/" + animationsFilePath);
        if (!file.exists()) {
            return;
        }

        try {
            JsonValue animationsRoot = new JsonReader().parse(file);
            JsonValue animations = animationsRoot.get("animations");
            if (animations == null || !animations.isArray()) {
                return;
            }

            for (JsonValue animation = animations.child; animation != null; animation = animation.next) {
                String id = animation.getString("id", null);
                String mediaFile = animation.getString("mediaFile", null);
                if (id == null || mediaFile == null || !looksLikeImageFile(mediaFile)) {
                    continue;
                }
                animationMediaById.put(id, mediaFile);
            }
        } catch (Exception ex) {
            Gdx.app.error("GameApp", "Failed to parse animations file", ex);
        }
    }

    private void collectAnimationMediaForLevel(JsonValue level, Array<String> levelImageFiles) {
        JsonValue sprites = level.get("sprites");
        if (sprites == null || !sprites.isArray()) {
            return;
        }

        for (JsonValue sprite = sprites.child; sprite != null; sprite = sprite.next) {
            String animationId = sprite.getString("animationId", null);
            if (animationId == null) {
                continue;
            }

            String mediaFile = animationMediaById.get(animationId);
            if (mediaFile != null && !levelImageFiles.contains(mediaFile, false)) {
                levelImageFiles.add(mediaFile);
            }
        }
    }

    private void collectImageFiles(JsonValue node, Array<String> output) {
        if (node == null) {
            return;
        }

        if (node.isArray()) {
            for (JsonValue value = node.child; value != null; value = value.next) {
                collectImageFiles(value, output);
            }
            return;
        }

        if (node.isObject()) {
            for (JsonValue child = node.child; child != null; child = child.next) {
                if (child.isString() && looksLikeImageField(child.name) && looksLikeImageFile(child.asString())) {
                    String relativePath = child.asString();
                    if (!output.contains(relativePath, false)) {
                        output.add(relativePath);
                    }
                }
                collectImageFiles(child, output);
            }
        }
    }

    private boolean looksLikeImageField(String fieldName) {
        if (fieldName == null) {
            return false;
        }
        return fieldName.endsWith("File");
    }

    private boolean looksLikeImageFile(String path) {
        if (path == null) {
            return false;
        }
        String normalized = path.toLowerCase();
        return normalized.endsWith(".png")
            || normalized.endsWith(".jpg")
            || normalized.endsWith(".jpeg")
            || normalized.endsWith(".bmp");
    }

    @Override
    public void dispose() {
        super.dispose();
        assetManager.dispose();
        batch.dispose();
        shapeRenderer.dispose();
        font.dispose();
    }
}

File: game_example_flutter/lib/game_app.dart
import 'dart:convert';

import 'package:flutter/services.dart';

import 'libgdx_compat/asset_manager.dart';
import 'libgdx_compat/game_framework.dart';
import 'libgdx_compat/gdx.dart';
import 'libgdx_compat/gdx_collections.dart';
import 'menu_screen.dart';

class GameApp extends Game {
  final AssetManager assetManager = AssetManager();
  final Array<String> menuOptions = Array<String>();
  final Array<String> levelNames = Array<String>();
  final Array<Array<String>> referencedImageFilesByLevel =
      Array<Array<String>>();
  final ObjectSet<String> queuedAssets = ObjectSet<String>();
  final ObjectMap<String, String> animationMediaById =
      ObjectMap<String, String>();
  final ObjectMap<String, String> animationGroupById =
      ObjectMap<String, String>();
  final ObjectMap<String, Array<String>> animationMediaByGroup =
      ObjectMap<String, Array<String>>();
  final Array<_AnimationMediaEntry> _animationMediaEntries =
      Array<_AnimationMediaEntry>();

  SpriteBatch? batch;
  ShapeRenderer? shapeRenderer;
  BitmapFont? font;

  Future<void> create() async {
    batch = SpriteBatch();
    shapeRenderer = ShapeRenderer();
    font = BitmapFont();
    font!.getData().markupEnabled = false;
    await _loadProjectData();
    setScreen(MenuScreen(this));
  }

  SpriteBatch getBatch() => batch!;

  ShapeRenderer getShapeRenderer() => shapeRenderer!;

  BitmapFont getFont() => font!;

  AssetManager getAssetManager() => assetManager;

  Array<String> getMenuOptions() => menuOptions;

  String getLevelName(int levelIndex) {
    if (levelIndex < 0 || levelIndex >= levelNames.size) {
      return 'Unknown';
    }
    return levelNames.get(levelIndex);
  }

  void queueReferencedAssetsForLevel(int levelIndex) {
    if (levelIndex < 0 || levelIndex >= referencedImageFilesByLevel.size) {
      return;
    }

    final Array<String> levelFiles = referencedImageFilesByLevel.get(
      levelIndex,
    );
    for (final String relativePath in levelFiles.iterable()) {
      final String assetPath = 'levels/$relativePath';
      if (!queuedAssets.contains(assetPath)) {
        assetManager.load(assetPath, Texture);
        queuedAssets.add(assetPath);
      }
    }
    assetManager.load('other/enrrere.png', Texture);
  }

  void unloadReferencedAssetsForLevel(int levelIndex) {
    if (levelIndex < 0 || levelIndex >= referencedImageFilesByLevel.size) {
      return;
    }

    final Array<String> levelFiles = referencedImageFilesByLevel.get(
      levelIndex,
    );
    for (final String relativePath in levelFiles.iterable()) {
      final String assetPath = 'levels/$relativePath';
      if (assetManager.isLoaded(assetPath, Texture)) {
        assetManager.unload(assetPath);
      }
      queuedAssets.remove(assetPath);
    }
  }

  Future<void> _loadProjectData() async {
    menuOptions.clear();
    levelNames.clear();
    referencedImageFilesByLevel.clear();
    animationMediaById.clear();
    animationGroupById.clear();
    animationMediaByGroup.clear();
    _animationMediaEntries.clear();

    try {
      final String gameDataRaw = await rootBundle.loadString(
        'assets/levels/game_data.json',
      );
      final Map<String, dynamic> root =
          jsonDecode(gameDataRaw) as Map<String, dynamic>;
      final List<dynamic>? levels = root['levels'] as List<dynamic>?;
      if (levels == null || levels.isEmpty) {
        _addFallbackLevels();
        return;
      }

      await _loadAnimationsFile(root);

      int index = 0;
      for (final dynamic rawLevel in levels) {
        final Map<String, dynamic>? level = rawLevel as Map<String, dynamic>?;
        if (level == null) {
          continue;
        }

        final String levelName = (level['name'] as String?) ?? 'Level $index';
        levelNames.add(levelName);
        menuOptions.add('LEVEL $index');
        final Array<String> levelImageFiles = Array<String>();
        _collectImageFiles(level, levelImageFiles);
        _collectAnimationMediaForLevel(level, levelImageFiles);
        referencedImageFilesByLevel.add(levelImageFiles);
        index++;
      }
    } catch (ex) {
      Gdx.app.error('GameApp', 'Failed to parse levels/game_data.json', ex);
      _addFallbackLevels();
    }
  }

  void _addFallbackLevels() {
    levelNames.add('Level 0');
    levelNames.add('Level 1');
    menuOptions.add('LEVEL 0');
    menuOptions.add('LEVEL 1');
    referencedImageFilesByLevel.add(Array<String>());
    referencedImageFilesByLevel.add(Array<String>());
  }

  Future<void> _loadAnimationsFile(Map<String, dynamic> root) async {
    final String? animationsFilePath = root['animationsFile'] as String?;
    if (animationsFilePath == null || animationsFilePath.isEmpty) {
      return;
    }

    try {
      final String raw = await rootBundle.loadString(
        'assets/levels/$animationsFilePath',
      );
      final Map<String, dynamic> animationsRoot =
          jsonDecode(raw) as Map<String, dynamic>;
      final List<dynamic>? animations =
          animationsRoot['animations'] as List<dynamic>?;
      if (animations == null) {
        return;
      }

      for (final dynamic rawAnimation in animations) {
        final Map<String, dynamic>? animation =
            rawAnimation as Map<String, dynamic>?;
        if (animation == null) {
          continue;
        }
        final String? id = animation['id'] as String?;
        final String? name = animation['name'] as String?;
        final String? mediaFile = animation['mediaFile'] as String?;
        final String groupId = (animation['groupId'] as String?) ?? '';
        if (id == null ||
            mediaFile == null ||
            !_looksLikeImageFile(mediaFile)) {
          continue;
        }

        animationMediaById.put(id, mediaFile);
        animationGroupById.put(id, groupId);

        if (groupId.isNotEmpty) {
          Array<String>? groupMedia = animationMediaByGroup.get(groupId);
          groupMedia ??= Array<String>();
          animationMediaByGroup.put(groupId, groupMedia);
          if (!groupMedia.contains(mediaFile, false)) {
            groupMedia.add(mediaFile);
          }
        }

        _animationMediaEntries.add(
          _AnimationMediaEntry(_normalize(name ?? ''), mediaFile),
        );
      }
    } catch (ex) {
      Gdx.app.error('GameApp', 'Failed to parse animations file', ex);
    }
  }

  void _collectAnimationMediaForLevel(
    Map<String, dynamic> level,
    Array<String> levelImageFiles,
  ) {
    final List<dynamic>? sprites = level['sprites'] as List<dynamic>?;
    if (sprites == null) {
      return;
    }

    final ObjectSet<String> spriteTokens = ObjectSet<String>();
    final ObjectSet<String> animationGroups = ObjectSet<String>();

    for (final dynamic rawSprite in sprites) {
      final Map<String, dynamic>? sprite = rawSprite as Map<String, dynamic>?;
      if (sprite == null) {
        continue;
      }

      final String animationId = (sprite['animationId'] as String?) ?? '';
      final String? mediaFile = animationMediaById.get(animationId);
      if (mediaFile != null && !levelImageFiles.contains(mediaFile, false)) {
        levelImageFiles.add(mediaFile);
      }

      final String? groupId = animationGroupById.get(animationId);
      if (groupId != null && groupId.isNotEmpty) {
        animationGroups.add(groupId);
      }

      _addTokens(spriteTokens, (sprite['type'] as String?) ?? '');
      _addTokens(spriteTokens, (sprite['name'] as String?) ?? '');
    }

    for (final String groupId in animationGroups.iterable()) {
      final Array<String>? groupMedia = animationMediaByGroup.get(groupId);
      if (groupMedia == null || groupMedia.size <= 0) {
        continue;
      }
      for (final String mediaFile in groupMedia.iterable()) {
        if (!levelImageFiles.contains(mediaFile, false)) {
          levelImageFiles.add(mediaFile);
        }
      }
    }

    if (spriteTokens.size <= 0 || _animationMediaEntries.size <= 0) {
      return;
    }

    for (final _AnimationMediaEntry entry
        in _animationMediaEntries.iterable()) {
      if (entry.normalizedName.isEmpty) {
        continue;
      }
      if (!_containsAnyToken(entry.normalizedName, spriteTokens)) {
        continue;
      }
      if (!levelImageFiles.contains(entry.mediaFile, false)) {
        levelImageFiles.add(entry.mediaFile);
      }
    }
  }

  void _collectImageFiles(dynamic node, Array<String> output) {
    if (node == null) {
      return;
    }

    if (node is List<dynamic>) {
      for (final dynamic value in node) {
        _collectImageFiles(value, output);
      }
      return;
    }

    if (node is Map<String, dynamic>) {
      for (final MapEntry<String, dynamic> entry in node.entries) {
        final String fieldName = entry.key;
        final dynamic value = entry.value;
        if (value is String &&
            _looksLikeImageField(fieldName) &&
            _looksLikeImageFile(value)) {
          if (!output.contains(value, false)) {
            output.add(value);
          }
        }
        _collectImageFiles(value, output);
      }
    }
  }

  bool _looksLikeImageField(String fieldName) {
    return fieldName.endsWith('File');
  }

  bool _looksLikeImageFile(String path) {
    final String normalized = path.toLowerCase();
    return normalized.endsWith('.png') ||
        normalized.endsWith('.jpg') ||
        normalized.endsWith('.jpeg') ||
        normalized.endsWith('.bmp');
  }

  bool _containsAnyToken(String normalizedValue, ObjectSet<String> tokens) {
    if (normalizedValue.isEmpty || tokens.size <= 0) {
      return false;
    }

    for (final String token in tokens.iterable()) {
      if (token.isNotEmpty && normalizedValue.contains(token)) {
        return true;
      }
    }
    return false;
  }

  void _addTokens(ObjectSet<String> tokens, String raw) {
    final String normalized = _normalize(raw);
    if (normalized.isEmpty) {
      return;
    }
    final List<String> split = normalized.split(RegExp(r'[^a-z0-9]+'));
    for (final String token in split) {
      if (token.length >= 3) {
        tokens.add(token);
      }
    }
  }

  String _normalize(String value) => value.trim().toLowerCase();

  @override
  void dispose() {
    super.dispose();
    assetManager.dispose();
    font?.dispose();
    shapeRenderer?.dispose();
  }
}

class _AnimationMediaEntry {
  final String normalizedName;
  final String mediaFile;

  _AnimationMediaEntry(this.normalizedName, this.mediaFile);
}
